analysis/src/cratesio.rs

```diff
@@ -0,0 +1,96 @@
+//! This module abstracts the communication with crates.io for a given crate
+//! Returns Error if the crate is not hosted on crates_io
+
+// TODO: A cheaper way to interact with crates.io can be working with their
+// experimental database dump that is updated daily, https://crates.io/data-access,
+// which will enable us to avoid making http requests and dealing with rate limits
+
+use anyhow::{anyhow, Result};
+use crates_io_api::SyncClient;
+use guppy::graph::PackageMetadata;
+use tabled::Tabled;
+
+#[derive(Tabled)]
+pub struct CratesioReport {
+    pub name: String,
+    pub is_hosted: bool,
+    pub downloads: u64,
+    pub dependents: u64,
+}
+
+pub struct CratesioAnalyzer {
+    client: SyncClient,
+}
+
+impl CratesioAnalyzer {
+    pub fn new() -> Result<Self> {
+        Ok(Self {
+            client: SyncClient::new(
+                "User-Agent: Whackadep (https://github.com/diem/whackadep)",
+                std::time::Duration::from_millis(1000),
+            )?,
+        })
+    }
+
+    pub fn analyze_cratesio(self, package: &PackageMetadata) -> Result<CratesioReport> {
+        let name = package.name();
+        let is_hosted = package.source().is_crates_io();
+
+        if !is_hosted {
+            return Err(anyhow!("Crate {} not hosted on crates.io", name));
+        }
```

Comment:
Instead of returning the error lets instead do an early return:

```
if package.source().is_crates_io() {
    // Filled in with 0 values
    return Ok(CratesioReport { ... });
}
```

This way the caller doesn't have to do the check ahead of time and it will simplify the logic for the caller.